massa-execution-worker/src/execution.rs

```diff
@@ -193,6 +193,68 @@ impl ExecutionState {
         self.active_history.write().0.push_back(exec_out);
     }
 
+    /// Helper function.
+    /// Within a locked execution context (lock is taken at the beginning of the function then released at the end):
+    /// - if not yet executed then transfer fee and add the operation to the context then return a context snapshot
+    ///
+    /// # Arguments
+    /// * `operation`: operation to be schedule
+    /// * `op_thread`: execution thread for the operation
+    /// * `sender_addr`: sender address for the operation (for fee transfer)
+
+    fn schedule_operation_for_execution(
+        &self,
+        operation: &SecureShareOperation,
+        op_thread: u8,
+        sender_addr: Address,
+    ) -> Result<ExecutionContextSnapshot, ExecutionError> {
+        let operation_id = operation.id;
+
+        // lock execution context
+        let mut context = context_guard!(self);
+
+        // ignore the operation if it was already executed
+        if context.is_op_executed(&operation_id) {
+            return Err(ExecutionError::IncludeOperationError(
+                "operation was executed previously".to_string(),
+            ));
+        }
+
+        // debit the fee from the operation sender
+        // fail execution if there are not enough coins
+        if let Err(err) =
+            context.transfer_coins(Some(sender_addr), None, operation.content.fee, false)
+        {
+            return Err(ExecutionError::IncludeOperationError(format!(
+                "could not spend fees: {}",
+                err
+            )));
+        }
+
+        // from here, the op is considered as executed (fees transferred)
+
+        // add operation to executed ops list
+        context.insert_executed_op(
+            operation_id,
+            None, // execution status not know yet
```

Comment:
Ok now I see why you have placed a None value. But I think you can remove the insert of the executed op here and add it only after the execution so that you already have the status. (Add it after the reset of the snapshot in case of error to avoid re-writing). It will allow you to remove the Option in the executed structures and will simplify a lot the rest of the code.
I don't see anywhere in the code where it could cause problems.